crates/ruff_linter/src/rules/flake8_pyi/rules/any_eq_ne_annotation.rs

```diff
@@ -78,23 +79,27 @@ pub(crate) fn any_eq_ne_annotation(checker: &mut Checker, name: &str, parameters
         return;
     }
 
-    if semantic.match_typing_expr(annotation, "Any") {
-        let mut diagnostic = Diagnostic::new(
-            AnyEqNeAnnotation {
-                method_name: name.to_string(),
-            },
-            annotation.range(),
-        );
-        // Ex) `def __eq__(self, obj: Any): ...`
-        diagnostic.try_set_fix(|| {
-            let (import_edit, binding) = checker.importer().get_or_import_builtin_symbol(
-                "object",
-                annotation.start(),
-                semantic,
-            )?;
-            let binding_edit = Edit::range_replacement(binding, annotation.range());
-            Ok(Fix::safe_edits(binding_edit, import_edit))
-        });
-        checker.diagnostics.push(diagnostic);
+    if !match_maybe_stringized_annotation(annotation, checker.locator(), |expr| {
```

Comment:
The downside of `match_maybe_stringized_annotation` is that we now end up re-parsing the same string-type annotation for every rule, which is somewhat expensive. 

We could instead:

* Make it a method on `Checker` or `SemanticModel` so that we could add caching if we wanted to (a simple cache could use a single slot)
* We could integrate the rule into `visit_deferred_string_type_definitions` and re-use the parsed type annotation. However, that would be a more invasive change

Reply:
> * We could integrate the rule into `visit_deferred_string_type_definitions`

I think that would be tricky because most of these `flake8-pyi` rules need to look at the surrounding function as a whole, rather than just a single annotation inside the function

Reply:
I centralized some of the logic into a method on `Checker` in https://github.com/astral-sh/ruff/pull/12951/commits/87a0c9c25b2dba76370e0139dc3fefa0d1c07aad so that it would be easy to add caching in the future